### PR TITLE
Add "final_extra_valid_opt_filepath" to train_model.py

### PR DESCRIPTION
**Pull request by moyapchen**
I've been annoyed at having to kick off separate eval loops at the ends of trains, so here we are. 

Also added a way to configure saving outputs to json (cause it's something that I've wanted for a while and I didn't want to add another return value in case scripts elsewhere were expecting only v_report and t_report). 

Not super attached to names or to the default values I've got set. I tried using 'Optional[str]' for the cmdline arg, but the test was not happy with that...

**Testing steps**
Wrote a new test that exercises all of the new functionality.

### CODE REVIEW

File: parlai/scripts/train_model.py [hunk 1]
```diff
@@ -87,6 +87,18 @@ def setup_args(parser=None) -> ParlaiParser:
         '--evaltask',
         help='task to use for valid/test (defaults to the one used for training)',
     )
+    train.add_argument(
+        '--final-extra-valid-opt-filepath',
```
Comment (stephenroller) [inline parlai/scripts/train_model.py:91]: Bike shedding but `--final-eval-opt`?

File: parlai/scripts/train_model.py [hunk 2]
```diff
@@ -87,6 +87,18 @@ def setup_args(parser=None) -> ParlaiParser:
...
+        type=str,
+        default='',
+        help="A '.opt' file that is used for final eval. Useful for setting skip-generation to false. 'datatype' must be included as part of the opt.",
+    )
+    train.add_argument(
+        '--write-log-as-json',
```
Comment (stephenroller) [inline parlai/scripts/train_model.py:97]: Aren't the final evals stored in the trainstats file already? If not, my suggestion would be we include them in the trainstats.

File: parlai/scripts/train_model.py [hunk 3]
```diff
@@ -919,6 +941,34 @@ def train(self):
 
         print_announcements(opt)
 
+        if opt['final_extra_valid_opt_filepath'] is not '':
```
Comment (stephenroller) [inline parlai/scripts/train_model.py:944]: Let's refactor this out plz

File: tests/test_train_model.py [hunk 4]
```diff
@@ -13,12 +13,69 @@
...
 from parlai.core.metrics import AverageMetric
 from parlai.core.worlds import create_task
+from parlai.core.opt import Opt
 from parlai.core.params import ParlaiParser
 from parlai.core.agents import register_agent, Agent
 from parlai.utils.data import DatatypeHelper
 
 
 class TestTrainModel(unittest.TestCase):
+    def test_final_extra_eval_and_save_json(self):
```
Comment (stephenroller) [inline tests/test_train_model.py:23]: Ty

File: parlai/scripts/train_model.py (hunk 2 above)
Comment (stephenroller) [reply parlai/scripts/train_model.py:97]: And not even make it an option. Just do it.

**Review by stephenroller — COMMENTED** [review]
Just want to finish discussion

File: parlai/scripts/train_model.py (hunk 1 above)
Comment (klshuster) [reply parlai/scripts/train_model.py:91]: +1

**Review by klshuster — COMMENTED** [review]
agree with all of @stephenroller's suggestions

File: parlai/scripts/train_model.py [hunk 5]
```diff
@@ -656,19 +650,43 @@ def _run_eval(
...
-                    json.dump(dict_report(report), f)
-            else:
-                with PathManager.open(
-                    opt['model_file'] + log_suffix + '.' + datatype, 'a'
-                ) as f:
-                    f.write(f'{metrics}\n')
+            with PathManager.open(
+                opt['model_file'] + log_suffix + '.' + datatype + ".json", 'a'
+            ) as f:
+                json.dump(dict_report(report), f)
```
Comment (stephenroller) [inline parlai/scripts/train_model.py:656]: I think we'd prefer keeping it with the rest of the `.trainstats`?

Comment (moyapchen) [reply parlai/scripts/train_model.py:656]: The other valid/reports aren't saved to .trainstats right now.

Comment (stephenroller) [reply parlai/scripts/train_model.py:656]: Every validation run, except the final one, is stored in trainstats.

https://github.com/facebookresearch/ParlAI/blob/9b5f8c670d21964ed140683a4e3fd1ae7e1f5db3/parlai/scripts/train_model.py#L474-L481

I'm humbly requesting you expand it to also include the final valid/test reports. (They can just be None during the intermediate stages)

Comment (moyapchen) [reply parlai/scripts/train_model.py]: Alright, wording wasn't quite clear there. Anyway, done.

**Review by stephenroller — APPROVED** [review]
Seems like you have a bug for when model file isn't set (and we should be throwing trainstats into the void). Otherwise lgtm, thanks for making the changes.

**Pull request merged**